src/view/treewalker.js

```diff
@@ -0,0 +1,438 @@
+	 *
+	 * @returns {Object} Object implementing iterator interface, returning information about taken step.
+	 */
+	next() {
+		if ( this.direction == 'FORWARD' ) {
+			return this._next();
+		} else {
+			return this._previous();
+		}
+	}
+
+	/**
+	 * Makes a step forward in view. Moves the {@link #position} to the next position and returns the encountered value.
+	 *
+	 * @private
+	 * @returns {Object}
+	 * @returns {Boolean} return.done True if iterator is done.
+	 * @returns {engine.view.TreeWalkerValue} return.value Information about taken step.
+	 */
+	_next() {
+		let position = Position.createFromPosition( this.position );
+		const previousPosition = this.position;
+		const parent = position.parent;
+
+		// We are at the end of the root.
+		if ( parent.parent === null && position.offset === parent.getChildCount() ) {
+			return { done: true };
+		}
+
+		// We reached the walker boundary.
+		if ( parent === this._boundaryEndParent && position.offset == this.boundaries.end.offset ) {
+			return { done: true };
+		}
+
+		// Get node just after current position.
+		let node;
+
+		// Text {@link engine.view.Text} element is a specific parent because it contains string instead of child nodes.
+		if ( parent instanceof Text ) {
+			node = parent._data[ position.offset ];
+		} else {
+			node = parent.getChild( position.offset );
+		}
+
+		if ( node instanceof Element ) {
+			if ( !this.shallow ) {
+				position = new Position( node, 0 );
+			} else {
+				position.offset++;
+			}
+
+			this.position = position;
+
+			return this._formatReturnValue( 'ELEMENT_START', node, previousPosition, position, 1 );
+		} else if ( node instanceof Text ) {
+			if ( this.singleCharacters ) {
+				position = new Position( node, 0 );
+				this.position = position;
+
+				return this._next();
+			} else {
+				let charactersCount = node._data.length;
+				let item = node;
+
+				// If text stick out of walker range, we need to cut it and wrap by TextProxy.
+				if ( node == this._boundaryEndParent ) {
+					charactersCount = this.boundaries.end.offset;
+					item = new TextProxy( node, 0, charactersCount );
+					position = Position.createAfter( item );
+				} else {
+					// If not just keep moving forward.
+					position.offset++;
+				}
+
+				this.position = position;
+
+				return this._formatReturnValue( 'TEXT', item, previousPosition, position, charactersCount );
+			}
+		} else if ( typeof node == 'string' ) {
+			let textLength;
+
+			if ( this.singleCharacters ) {
+				textLength = 1;
+			} else {
+				// Check if text stick out of walker range.
+				const endOffset = parent === this._boundaryEndParent ? this.boundaries.end.offset : parent._data.length;
+
+				textLength = endOffset - position.offset;
+			}
+
+			const textProxy = new TextProxy( parent, position.offset, textLength );
+
+			position.offset += textLength;
+
+			// Position at the end of Text is always out of Text node not inside.
+			if ( position.offset == parent._data.length ) {
```

Comment:
I think I should move this condition here: https://github.com/ckeditor/ckeditor5-engine/pull/492/files#diff-d0dfec7faddd4ddd3ee23c8318b99147R384